### ISSUE TITLE
Share same pre-embedding file in different token_embedders

### ISSUE THREAD

**Issue by valdersoul** [allenai/allennlp#348]
Hi, I want to share same pre-trained embedding file e.t glove, while with two different token_embedders (because they are from different text_field). Is there any way to share this file between them?

Thanks for your attention.

**Comment by valdersoul** [comment]
Or we can make the shared file as a value for all instances of the Embedding class.

**Comment by DeNeutoy** [comment]
I am slightly unsure of your question. If you are looking to use the same embedding (i.e the same model parameters), you can just use the same text field embedder for both `TextFields` and use the same namespace for both `TextFields`. 

If you want to have two separate sets of parameters which are initialized using the same pre-trained vectors, you can just have two separate `TokenEmbedders`, both of which are embeddings.

Example:
```
   // These have different parameters, but are initialised using the same pre-trained vectors.
    "text_field_embedder1": {
      "key_of_the_token_indexer_for_text_field_1": {
        "type": "embedding",
        "pretrained_file": "/glove/glove.6B.100d.txt.gz",
        "embedding_dim": 100,
        "trainable": false
      }
},
   "text_field_embedder2": {
      "key_of_the_token_indexer_for_text_field_2": {
        "type": "embedding",
        "pretrained_file": "same/path/to/glove/vectors",
        "embedding_dim": 100
        }
}
```
If you want to share the embedding, just use the same namespace for both text fields and have a single `TextFieldEmbedder` which you apply twice in your model (and therefore only have one of these text_field_embedders in the JSON config above). Hopefully one of these is your use case!

You are however correct that this results in reading the glove vectors in twice, which is unfortunate - we hadn't run in to this issue. Let us know if it really slows you down, e.g you want to have many namespaces which differ but share glove vectors. Hopefully this solves your problem in the short term.

**Comment by valdersoul** [comment]
Thanks for your reply. In my situation, is the second one. Yes, I can do this by two text_field_embedder, I just wonder whether there is a better way. I'll close this one.

**Comment by matt-gardner** [comment]
Are you sure you want to have two different `TextFieldEmbedders`?  In almost any NLP model, you probably just need one of those.  For BiDAF, for instance, we use the same embedder on both the question and the passage, which are both text fields.  What use case do you have where you really want to have two different ones?

**Comment by valdersoul** [comment]
For example, same words with different structures (sentence, paragraph), they need different paddings. It's easier to do in the data-prepare stage than in the model level.

**Comment by matt-gardner** [comment]
Yeah, you don't need separate embeddings for this.  In SQuAD, we have a paragraph and a question that receive different padding lengths, but share their embedding layers.  You just create two different `TextField` objects, with difference keys in the instance, and padding is done _per key_ in the `Instance` object.

We build the `Instance` objects in the dataset reader, with the relevant lines here:

https://github.com/allenai/allennlp/blob/b12517c78db605eda815a0468fd92c014e37821b/allennlp/data/dataset_readers/squad.py#L173-L175

Then, in the model code, we just use the same `TextFieldEmbedder` to embed the two different fields:

https://github.com/allenai/allennlp/blob/b12517c78db605eda815a0468fd92c014e37821b/allennlp/models/bidaf.py#L186-L187

I'm pretty sure the code already does what you want it to.  The only time you'd want two _different_ embedders is if you wanted to, e.g., use character-level information for the question, but not the passage.  We could in principle do this, but sharing the embedder is currently not possible when building things `from_params`.

**Comment by valdersoul** [comment]
Yes, in this situation, the code works well. How about this situation? A paragraph is as below:

a b d f d? f d t! i o j s. 

The paragraph level can be done as before. If we need to pad the sentence into:
a b d f d
f d t 0 0
i o j s 0.

This needs to be added as a separate token_indexers. 
Correct me if I'm wrong.

**Comment by matt-gardner** [comment]
In this situation, you'd want a `ListField` holding a list of `TextFields`.  You break up the paragraph into sentences yourself, in your `DatasetReader`, make `TextFields` with each of them (using the same `TokenIndexers`), then put them into a `ListField`.  In your model code, this will give you an array of shape `(batch_size, num_sentences, sentence_length)` for the basic word-id representation, and you should be able to pass this into a `TextFieldEmbedder` without issue.  We haven't actually tested this with any of our models, though, so it's possible there's still a bug somewhere in this use case - you might have to `TimeDistribute` the `TextFieldEmbedder`, for instance.

**Comment by valdersoul** [comment]
I‘ll have a try. Thanks for your reply!
